Allow to query for passwords in CowrieSession API - Closes #607

IshaanXCoder opened the pull request:
# Description
This PR includes the fixes discussed in https://github.com/intelowlproject/GreedyBear/pull/611 . This adds the ability to query the CowrieSession API by **password**, in addition to the existing **IP address** and **SHA-256 hash** query methods. Changes are mentioned as follows - 

- `api/views/cowrie_session.py`
  - Added password query support using Django ORM (removed the raw SQL added in the previous PR).
  - Simplified query validation: IP → hash → password fallback.
  - Removed character restrictions and redundant IP/hash re-checks as asked.

- `greedybear/models.py`
  - Added `GinIndex` on `CowrieSession.credentials` in `Meta.indexes` to speed up credential lookups.
  - Removed the manual migration done earlier

I also added .DS_Store to .gitignore since it can be irritating for developers.

## Related issues
SOLVES #607 

## Type of change
- [x] New feature (non-breaking change which adds functionality).


# Checklist
  
- [x] I have read and understood the rules about how to contribute to this project.
- [x] The pull request is for the branch `develop`. 
- [x] I have added documentation of the new features. (Updated docstring in `cowrie_session_view`)
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [x] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check `CONTRIBUTE.md`). (Updated view docstring)
- [ ] If the GUI has been modified:
  - [ ] I have provided a screenshot of the result in the PR.
  - [ ] I have created new frontend tests for the new component or updated existing ones.

IshaanXCoder commented:
> Please fix the formatting @IshaanXCoder

Made the required changes

regulartim commented:
>  Made the required changes

No you did not. Since I asked you to fix the formatting you did not push a single commit!

regulartim commented:
CI still fails. Please read into the contribution guidelines again, learn how to avoid such problems and make sure your code is correctly formatted before pushing.

IshaanXCoder commented:
Hey @regulartim PTAL. I fixed it, took time, apologies for the same.
 
- Added input validation to properly distinguish between query types:
  - Invalid IPs/hashes return 400 (Bad Request) instead of being treated as password queries.
  - XSS attempts and malformed input are rejected.
  - Only well-formed, non-IP/non-hash strings are processed as password queries
- Fixed formatting and linting issues using black, isort and flake8, all passed. Understood where I was going wrong earlier. 

<img width="672" height="107" alt="Screenshot 2025-12-24 at 3 22 28 PM" src="https://github.com/user-attachments/assets/07a175f5-86f7-4984-8e16-bc0a56d7a7a2" />
<img width="694" height="181" alt="Screenshot 2025-12-24 at 3 22 51 PM" src="https://github.com/user-attachments/assets/92083a5e-27c9-4faa-8e96-7506ae36e0a0" />

regulartim commented on greedybear/migrations/0001_initial.py:
Why did you touch all of these old migration files?

IshaanXCoder replied:
<img width="773" height="232" alt="image" src="https://github.com/user-attachments/assets/d45f9fa7-69df-4906-9a63-95cdbb3b8906" />

The files were just reformatted, nothing has been changed with the code. If you want i'll change it back to the original format.

regulartim replied:
Ok, strange. Please only include files in the PR that are actually relevant to the issue that is addressed. And old migrations are not.

regulartim commented on api/views/cowrie_session.py:
```diff
@@ -74,19 +78,32 @@ def cowrie_session_view(request):
         return HttpResponseBadRequest("Missing required 'query' parameter")
 
+    if "<" in observable or ">" in observable or observable.count("}") > observable.count("{"):
+        return HttpResponseBadRequest("Invalid query parameter: contains invalid characters")
```
Why are these chars invalid? Couldn't they be part of a password?

IshaanXCoder replied:
Earlier i was thinking in the wrong direction, I was thinking around XSS protection, but it is not an issue here , will remove this entirely.

regulartim replied:
Ok

regulartim commented on api/views/cowrie_session.py:
```diff
@@ -74,19 +78,32 @@ def cowrie_session_view(request):
 
+    if "<" in observable or ">" in observable or observable.count("}") > observable.count("{"):
+        return HttpResponseBadRequest("Invalid query parameter: contains invalid characters")
+    looks_like_ip = "." in observable or ":" in observable or "/" in observable
```
Why does the observable look like an IP when it contains a "/"?

IshaanXCoder replied:
/ is used in CIDR notations. 

<img width="987" height="130" alt="Screenshot 2025-12-24 at 6 50 30 PM" src="https://github.com/user-attachments/assets/e63d5f0f-6b75-40cd-b8af-587c8a39ca06" />

reference - https://en.wikipedia.org/wiki/IP_address

regulartim commented on api/views/cowrie_session.py:
```diff
@@ -74,19 +78,32 @@ def cowrie_session_view(request):
     else:
-        return HttpResponseBadRequest("Query must be a valid IP address or SHA-256 hash")
+        password_pattern = f" | {observable}"
+        sessions = CowrieSession.objects.filter(duration__gt=0, credentials__icontains=password_pattern).prefetch_related("source", "commands")
```
Why did you choose to use icontains instead of contains here? Passwords are usually case sensitive, right?

regulartim replied:
But I you wrote tests for that and they pass, right? Where did you get the idea to use `filter` like that?

IshaanXCoder replied:
It worked in practice on PostgreSQL and that’s where the idea came from, not from the docs.

I agree this being undocumented and  relies on implicit casting, is not  a good approach. I'll fix it accordingly.

regulartim replied:
Ok :)

regulartim commented on greedybear/models.py:
```diff
@@ -90,6 +101,7 @@ class CowrieSession(models.Model):
     class Meta:
         indexes = [
             models.Index(fields=["source"]),
+            GinIndex(fields=["credentials"], name="greedybear_credentials_gin_idx"),
```
Are you sure that this index helps in our specific use case? I looked it up and that's really advanced database stuff! At least for me it is hard to understand what this type of index actually does.

IshaanXCoder replied:
I don’t think this GIN index actually helps in its current form. We only query credentials via substring matching and a plain GIN index on an ArrayField won’t be used for that by PostgreSQL.
unless you have a concrete query in mind that would benefit from array containment, I’ll remove the index to avoid extra complexity

IshaanXCoder replied:
If we switch from substring matching to exact credential matches, PostgreSQL can use a GIN index efficiently and avoid full scans. This would help us to scale better. 

But in my opinion, if we see in long term, a better approach would be :
Stop storing credentials as an array inside CowrieSession, instead, store each credential as its own row in a separate table. This'll avoid scanning every session

regulartim replied:
> If we switch from substring matching to exact credential matches, PostgreSQL can use a GIN index efficiently and avoid full scans. This would help us to scale better.

But that's not possible, because we only query passwords, not credentials, right?

regulartim replied:
> But in my opinion, if we see in long term, a better approach would be :
Stop storing credentials as an array inside CowrieSession, instead, store each credential as its own row in a separate table. This'll avoid scanning every session

I agree. Would you like to implement that?

IshaanXCoder replied:
Yeah sure, I would love to work on this and try to implement this at the earliest. I'll create a new issue for the same.

regulartim reviewed (changes requested):
No problem, take your time! :) However, in my opinion, this is not ready to be merged yet. We have to discuss some things.

IshaanXCoder commented:
@regulartim shall i implement the above conversation?
